feat(comments): support prefix comments on Record's key-value pairs.

german1608 opened the pull request:
This PR tries to add support to record field comments both on formatting and dhall-docs documentation extension. Specifically the following ones:

```dhall
{
  -- before each record field
  a : Text
  -- after each record field
, ...
}
```

I updated the `Record`'s `Expr` constructor from:

```haskell
Record (Map Text (Expr s a))
```

to

```haskell
Record (Map Text (RecordField s a))
```

where `RecordField` stores the last of the expression on the right side of the `:` and the prefix and suffix comments. Although the prefix comment is annotating the label field, the best place to save that comment is there.

However, I'm having a problem with `unsafeSubExpressions`, because it can't traverse the `Maybe s` I put on `RecordField`. A quick solution for that is to change those `RecordField`'s `Maybe s`' to `Nothing`, although that means we loose that information when traversing.

I'll keep this PR as a draft to get your opinion in that matter.

sjakobi commented:
I'm happy to see that you've started working on this! :)

> This PR tries to add support to record field comments both on formatting and dhall-docs documentation extension. Specifically the following ones:
> 
> ```haskell
> {
>   -- before each record field
>   a : Text
>   -- after each record field
> , ...
> }
> ```

Have you considered including the whitespace between the label and the value, e.g. `{ x {- A -} : {- B -} T }`? These might not be great spots for documentation comments, but in the context of #145, it would probably be useful anyway.

> However, I'm having a problem with `unsafeSubExpressions`, because it can't traverse the `Maybe s` I put on `RecordField`. A quick solution for that is to change those `RecordField`'s `Maybe s`' to `Nothing`, although that means we loose that information when traversing.

I haven't checked, but I think we should be able to preserve the comments here. We have a similar case with the `Binding` in `Let`. Try copying `bindingExprs`!

If you need more help, please shout! :)

sjakobi commented on dhall/src/Dhall/Syntax.hs:
```diff
@@ -284,6 +286,31 @@ instance Bifunctor PreferAnnotation where
+-}
+data RecordField s a = RecordField
+    { recordFieldSrc0 :: Maybe s
+    , field :: Expr s a
```
To me, the "field" is the combination of the label and the value.

```suggestion
    , recordFieldValue :: Expr s a
```

german1608 commented:
> Have you considered including the whitespace between the label and the value, e.g. `{ x {- A -} : {- B -} T }`? These might not be great spots for documentation comments, but in the context of #145, it would probably be useful anyway.

I didn't consider them because as you say is an awkward place to write documentation. But I can add it with no problem :)

> I haven't checked, but I think we should be able to preserve the comments here. We have a similar case with the `Binding` in `Let`. Try copying `bindingExprs`!

I haven't check out `bindingExprs` yet, but on `unsafeSubExpressions` the `Let` constructor is handled with the `unhandledConstructor`.

I'll read `bindingExprs` to see what can I do to work further.

> If you need more help, please shout! :)

Cool! What do you think might suffix for tests to this? A format test?

I think that to make this PR not too big I'll add the record field support for `dhall-docs` on another PR.

sjakobi commented:
> I haven't check out `bindingExprs` yet, but on `unsafeSubExpressions` the `Let` constructor is handled with the `unhandledConstructor`.

Oh, right! I forgot why `unsafeSubExpressions` is separate! I guess we should move the `Record` handling from `unsafeSubExpressions` to `subExpressions` then. We'll also need to update the other places where we currently use `unsafeSubExpressions`.

> Cool! What do you think might suffix for tests to this? A format test?

Take a look at the tests added in https://github.com/dhall-lang/dhall-haskell/pull/1273.

Formatting idempotence is a bit tricky to get right, but we can fix it later if necessary. We can extend some property tests to help with that: https://github.com/dhall-lang/dhall-haskell/issues/1465.

sjakobi commented:
BTW, you could try making the same changes for `RecordLit` in this PR – I believe that the changes would look very similar to those for `Record`.

Documentation comments on record literals could also prove valuable, for example in `package.dhall` files, or on schema records.

german1608 commented:
> Documentation comments on record literals could also prove valuable, for example in `package.dhall` files, or on schema records.

You're right! After getting record types right I'll do it.

Looks like I have a lot of work to do ;)

Thanks for your guidance @sjakobi

46511dedi commented:
,😉

german1608 commented:
I successfully captured the prefix whitespace of a key-value pair, but the suffix is consumed by the annotation expression. Any idea about how to solve this?

For example,

```dhall
{ -- foo
  a : Text
  -- baz
}
```

results in (approximate)

```
...
("a",
RecordField
  { recordFieldSrc0 = " -- foo"
  , recordFieldValue = Note (... "Text \n  -- baz") Text
  , recordFieldSrc1 = ""
  }
)
```

I believe that almost all expressions consume the suffix whitespace around it

Any ideas? cc: @Gabriel439 @sjakobi

sjakobi commented:
> I successfully captured the prefix whitespace of a key-value pair, but the suffix is consumed by the annotation expression. Any idea about how to solve this?

Oh yes. I forgot that this is a constraint that we also encountered with `let` in https://github.com/dhall-lang/dhall-haskell/pull/1273.

I guess we could try fiddling with the `Note`s of the type, but even if it would work, it would probably be rather messy.

I hope @Gabriel439 has a better idea – he definitely knows more about parsing!

Gabriel439 commented:
@german1608: There isn't a good solution to this and the `let`-related comment parsing has the same limitation.  The only way to solve this "for real" is to preserve all whitespace/comments for all constructors in the syntax tree.  For now, it's fine to not preserve that whitespace

german1608 commented:
I'll post an update.

I dropped `recordFieldSrc1` (the suffix comment) because of what we last discussed. After that, I updated `RecordLit` Map to use `RecordField` as map values.

I started the pretty printing of those comments, this is what I have at this moment:

With this file:
```dhall
{ -- hi there
  -- testing around
  a: Text,

  {- foo
  -}
  b: Bool
}
```

`dhall format` will return:
```dhall
{ -- hi there
  -- testing around
  a :
    Text
,

  {- foo
  -}
  b :
    Bool
}
```

for me, it's awkward that the type follows in a newline. Do you know the reason? I'm not too fluent on `prettyprint` :)

Also, what other test cases you think will be helpful for this? something similar to let-bindings as we had on #1273?

I'll keep these answer open and debug further tomorrow.

german1608 commented:
These line is what causes the line break there:

https://github.com/dhall-lang/dhall-haskell/blob/71e485bc87bd7fed3ba72e014bf79429acdf4056/dhall/src/Dhall/Pretty/Internal.hs#L1317-L1319

Changing these lines to:

```haskell
space <> prettyValue val
```

and supplying a file like this:

```dhall
{ -- hi there
  -- testing around
  a: Text,
  {- foo
  -}
  b: Bool
}
```

outputs a file like this:

```dhall
{ -- hi there
  -- testing around
  a : Text
,
  {- foo
  -}
  b : Bool
}
```

What do you think of this formatting?

german1608 commented:
I believe that this is ready for review. You can check out the test-cases I added. What I haven't done yet is:

* Add support for other comments i.e. `a {- B -} : {- C -} T` because I'm unsure how to format them properly yet. Also, I'd like to finish this PR as soon as possible to work again at `dhall-docs` relevant features.
* Check idempotent tests setup. I'll try to address #1465 on this PR

I improved the formatting of key-value pairs to avoid the newline after the `:` or `=` as I showed in this [comment](https://github.com/dhall-lang/dhall-haskell/pull/1908#issuecomment-658546311). The relevant code is here:

https://github.com/dhall-lang/dhall-haskell/blob/79c00ae1712acf49223ff7522af1bf3226e63ecd/dhall/src/Dhall/Pretty/Internal.hs#L1333-L1374

I'll work to improve the idempotent tests before marking this PR as ready for review, but you can take a look if you want to

sjakobi commented:
Apologies for the delay @german1608! I've started reviewing this now.

Unfortunately the mass of style changes make the review much harder. I wish I had remembered to suggest that you do a complete `stylish-haskell` sweep over the code base before doing this work. :(

Could you possibly still do that and then rebase this PR?

german1608 commented:
> Apologies for the delay @german1608! I've started reviewing this now.

No worries!

> Unfortunately the mass of style changes make the review much harder. I wish I had remembered to suggest that you do a complete `stylish-haskell` sweep over the code base before doing this work. :(
> 
> Could you possibly still do that and then rebase this PR?

Sure thing! Give me 30minutes do the change.


> My suggestion is:
> 
> * Backup this work on a different branch.
> * Remove the work on the pretty-printing logic from this PR, as if the `recordFieldSrc` didn't exist.
> * I will give the remaining code a speedy review.
> * Once it's merged you can focus on dhall-docs again, and for example work on using the record field comments in `dhall-docs`.

> Does this sound like a reasonable plan?

Yeah, let's do it.

I'll ping you when I'm done @sjakobi

german1608 commented:
@sjakobi the stylish-haskell changes you say are only on imports formatting, right? Aside that, I manually changed in some places:

```haskell
do e
```

to

```haskell
e
```

when it's valid to do so. Should I do these changes on another PR as well?

sjakobi commented:
> Aside that, I manually changed in some places:
> 
> ```haskell
> do e
> ```
> 
> to
> 
> ```haskell
> e
> ```
> 
> when it's valid to do so. Should I do these changes on another PR as well?

I don't feel very strongly about `do e` vs. `e` – I mostly don't like the added noise of these changes.

Maybe just do what seems "economical" for your time and the reviewers' time.

german1608 commented:
Ok, I'll revert those changes too

german1608 commented:
additions/deletions before stylish-haskell changes: +1,399 −758

additions/deletion after reverting them: +460 −270 

I believe that you can review now, @sjakobi

german1608 commented on dhall/src/Dhall/Syntax.hs:
```diff
@@ -284,6 +287,41 @@ instance Bifunctor PreferAnnotation where
+this is the best place to keep these comments
+-}
+data RecordField s a = RecordField
+    { recordFieldSrc0  :: Maybe s
```
I'm missing this.

german1608 commented on dhall/src/Dhall/Syntax.hs:
```diff
@@ -514,20 +563,27 @@ instance Applicative (Expr s) where
-    expression >>= k = Lens.over unsafeSubExpressions (>>= k) expression
+    Note a b    >>= k = Note a (b >>= k)
+    expression  >>= k = case expression of
+        Record a -> Record $ f <$> a
+        RecordLit a -> RecordLit $ f <$> a
```
The reason I created another case for them because I reuse `f` on both and I didn't want to put `f` on the global level

german1608 commented on dhall/src/Dhall/Syntax.hs:
```diff
@@ -708,6 +766,17 @@ bindingExprs f (Binding s0 n s1 t s2 v) =
 
+{-| Traverse over the immediate 'Expr' children in a 'RecordField'.
+-}
+recordFieldExprs
```
I put it there because the other `*Exprs` traversal functions (i.e. `bindingExprs`, `chunkExprs`, etc) are on that section as well.

sjakobi commented:
Regarding the style changes, do you want to put them up in a separate PR?

german1608 commented:
> Regarding the style changes, do you want to put them up in a separate PR?

Yes, I believe that it will make future contributions shorter and users of stylish-haskell won't have to get through this :)

I'll make the PR today soon

sjakobi commented on dhall/src/Dhall/TypeCheck.hs:
```diff
@@ -747,7 +748,7 @@ infer typer = loop
                     return _T'
 
-            xTs <- traverse process (Dhall.Map.sort xts)
+            xTs <- traverse process (Dhall.Map.sort $ recordFieldValue <$> xts)
```
Better move the `recordFieldValue` into `process` above.

german1608 commented on dhall/src/Dhall.hs:
```diff
@@ -1091,15 +1091,17 @@ pairFromMapEntry :: Decoder k -> Decoder v -> Decoder (k, v)
     extractOut (RecordLit kvs)
-        | Just key <- Dhall.Map.lookup "mapKey" kvs
-        , Just value <- Dhall.Map.lookup "mapValue" kvs
+        | Just key <- Dhall.Core.recordFieldValue <$> Dhall.Map.lookup "mapKey" kvs
```
I imported `Dhall.Core` as `Core` as you suggested on this [comment](https://github.com/dhall-lang/dhall-haskell/pull/1908#discussion_r455815593)

german1608 commented on dhall/src/Dhall.hs:
```diff
@@ -1091,15 +1091,17 @@ pairFromMapEntry :: Decoder k -> Decoder v -> Decoder (k, v)
         k' <- expected k
         v' <- expected v
-        pure $ Record $ Dhall.Map.fromList [("mapKey", k'), ("mapValue", v')]
+        pure $ Record $ Dhall.Core.makeRecordField <$> Dhall.Map.fromList
```
It looks better actually. Also, there are other modules where `Core` is used instead of `Dhall.Core`

german1608 commented on dhall/src/Dhall/Parser/Expression.hs:
```diff
@@ -474,14 +474,10 @@ parsers embedded = Parsers {..}
             alternative04 = (do
```
The reason I add those `try`s is documented on the source code, but it's because to not loose the `whitespace` consumed.

I'm not to fluent on parser combinators, probably it could be improved

sjakobi replied:
Thanks for the explanations!

I think we should do some "left-factoring" here, which IIUC in this case means that we try to parse the first whitespace of a record (type) literal in a central place and then pass it to the following parsers.

@Gabriel439 can probably explain this more clearly.

I think we can address this in a follow-up PR though.

Gabriel439 replied:
Yeah, these `try`s can lead to large parsing performance regressions, so it's worth seeing if we can narrow their scope

german1608 commented:
> Great job @german1608! This clearly was a lot of onerous work!
> 
> My inline comments are mostly about style or performance.
> 
> I'm quite curious how much this affects performance at this stage. Not sure what would be a good benchmark… Maybe `dhall-to-yaml` one of the `dhall-kubernetes` examples?!

Here are the results of my tests:

| Example                             | Without this change                                                  | With this change                                                     |
|-------------------------------------|----------------------------------------------------------------------|----------------------------------------------------------------------|
| `aws-iam-authenticator-chart.dhall` | `real    0m0,237s`<br> `user    0m0,206s`<br> `sys     0m0,031s`<br> | `real    0m0,301s`<br> `user    0m0,278s`<br> `sys     0m0,018s`<br> |
| `deployment.dhall`                  | `real    0m0,236s`<br> `user    0m0,231s`<br> `sys     0m0,005s`<br> | `real    0m0,329s`<br> `user    0m0,309s`<br> `sys     0m0,016s`     |
| `deploymentSimple.dhall`            | `real    0m0,235s`<br> `user    0m0,206s`<br> `sys     0m0,029s`<br> | `real    0m0,243s`<br> `user    0m0,210s`<br> `sys     0m0,032s`<br> |
| `ingress.dhall`                     | `real    0m0,261s`<br> `user    0m0,245s`<br> `sys     0m0,013s`<br> | `real    0m0,248s`<br> `user    0m0,233s`<br> `sys     0m0,015s`<br> |
| `service.dhall`                     | `real    0m0,412s`<br> `user    0m0,323s`<br> `sys     0m0,027s`<br> | `real    0m0,247s`<br> `user    0m0,238s`<br> `sys     0m0,008s`<br> |

While the numbers are greater with this change, it isn't a **big** difference

german1608 commented on dhall-json/src/Dhall/JSON.hs:
```diff
@@ -966,12 +973,14 @@ convertToHomogeneousMaps (Conversion {..}) e0 = loop (Core.normalize e0)
             Core.Record a'
           where
-            a' = fmap loop a
+            f (Core.RecordField s0 re) = Core.RecordField s0 (loop re)
+            a' = fmap f a
```
It works if I use `Lens.over`:

```suggestion
            a' = Lens.over Core.recordFieldExprs loop <$> a
```

What do you think? @sjakobi

german1608 replied:
What I don't like of that approach is that we're adding `lens-family-core` to `dhall-json.cabal` to just do that.

sjakobi replied:
@german1608 That looks like a good solution to me. Adding `lens-family-core` as a direct dependency isn't a problem – that package is already a transitive dependency via `dhall`.

If we can make the "non-core" packages less dependent on the exact shape of `Expr`, future refactorings will be easier.

german1608 replied:
Cool! I've pushed that change :smile:

german1608 commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -1305,3 +1310,4 @@ alphaNormalize = goEnv EmptyNames
         go                     = goEnv e0
         goBind x               = goEnv (Bind e0 x)
         goChunks (Chunks ts x) = Chunks (fmap (fmap go) ts) x
+        goRecordField (RecordField s0 e) = RecordField s0 (go e)
```
`Lens.over` does the trick here :sunglasses:

sjakobi commented on dhall/src/Dhall/Eval.hs:
```diff
@@ -1273,9 +1274,9 @@ alphaNormalize = goEnv EmptyNames
+                Record (Lens.over Syntax.recordFieldExprs go <$> kts)
             RecordLit kts ->
-                RecordLit (goRecordField <$> kts)
+                RecordLit (Lens.over Syntax.recordFieldExprs go <$> kts)
```
Actually the previous version seems clearer. This version has more "non-essential" complexity that we're trying to avoid in the core modules.

 Could you revert efb4cfde2f766b3d8b5a0fffeb9af65116dec1a0?

sjakobi commented:
Regarding the benchmarks: If the time for `deployment.dhall` really went up by ~50%, that doesn't bode well for larger jobs.

Since the cost of handling the record fields is so spread out, I don't see an obvious way to improve the performance again.

BTW, there's a nice tool that you can use to produce more accurate timings of CLI programs: http://hackage.haskell.org/package/bench

Gabriel439 commented:
Give me some time to review the parsing code.  I can help with the left factoring

Gabriel439 commented on dhall/src/Dhall/Parser/Expression.hs:
```diff
@@ -474,14 +474,12 @@ parsers embedded = Parsers {..}
-                whitespace
-
-                _ <- optional (_comma *> whitespace)
+                -- If it fails to parse the _comma we need to recover the
```
Instead of wrapping these all in `try`, I think you can parse the whitespace once (wrapped in `src`), and pass the parsed `Src` span as a function argument to each of the alternative record parses.  I'm fairly confident that will be more efficient

german1608 commented:
@sjakobi I used `bench` as you suggested and these are the new reports:

Running on `master`:

```
benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/aws-iam-authenticator-chart.dhall
time                 250.0 ms   (-251.3 ms .. 764.7 ms)
                     0.645 R²   (0.071 R² .. 1.000 R²)
mean                 364.4 ms   (290.4 ms .. 489.3 ms)
std dev              121.4 ms   (15.56 ms .. 153.1 ms)
variance introduced by outliers: 73% (severely inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/deployment.dhall
time                 215.7 ms   (160.6 ms .. 241.7 ms)
                     0.984 R²   (0.940 R² .. 1.000 R²)
mean                 251.1 ms   (234.0 ms .. 275.4 ms)
std dev              26.13 ms   (10.80 ms .. 34.99 ms)
variance introduced by outliers: 19% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/deploymentSimple.dhall
time                 256.7 ms   (216.3 ms .. 312.4 ms)
                     0.987 R²   (0.973 R² .. 1.000 R²)
mean                 236.6 ms   (227.7 ms .. 249.4 ms)
std dev              13.70 ms   (6.742 ms .. 17.98 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/ingress.dhall
time                 240.0 ms   (219.0 ms .. 260.5 ms)
                     0.997 R²   (0.991 R² .. 1.000 R²)
mean                 227.8 ms   (222.7 ms .. 234.2 ms)
std dev              7.514 ms   (3.542 ms .. 8.519 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/service.dhall
time                 257.0 ms   (228.0 ms .. 280.8 ms)
                     0.993 R²   (0.984 R² .. 1.000 R²)
mean                 229.9 ms   (220.6 ms .. 241.6 ms)
std dev              15.30 ms   (9.510 ms .. 21.69 ms)
variance introduced by outliers: 15% (moderately inflated)
```

Running on this PR without the "left-refactoring" improvement:

```
benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/aws-iam-authenticator-chart.dhall
time                 285.2 ms   (244.9 ms .. 323.2 ms)
                     0.994 R²   (0.991 R² .. 1.000 R²)
mean                 256.6 ms   (245.6 ms .. 269.1 ms)
std dev              14.80 ms   (8.708 ms .. 20.20 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/deployment.dhall
time                 255.0 ms   (248.6 ms .. 260.5 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 245.1 ms   (232.4 ms .. 250.0 ms)
std dev              9.991 ms   (426.1 μs .. 12.89 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/deploymentSimple.dhall
time                 236.8 ms   (228.4 ms .. 252.0 ms)
                     0.998 R²   (0.994 R² .. 1.000 R²)
mean                 237.4 ms   (233.8 ms .. 241.3 ms)
std dev              5.500 ms   (3.970 ms .. 6.648 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/ingress.dhall
time                 280.1 ms   (238.9 ms .. 350.2 ms)
                     0.981 R²   (0.959 R² .. 1.000 R²)
mean                 256.3 ms   (247.0 ms .. 274.1 ms)
std dev              16.45 ms   (5.208 ms .. 21.33 ms)
variance introduced by outliers: 17% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/service.dhall
time                 363.2 ms   (262.9 ms .. 500.4 ms)
                     0.952 R²   (0.929 R² .. 1.000 R²)
mean                 274.2 ms   (249.6 ms .. 320.8 ms)
std dev              45.81 ms   (9.337 ms .. 64.70 ms)
variance introduced by outliers: 38% (moderately inflated)
```

These are the bench results with the "left-refactoring" technique:

```
benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/aws-iam-authenticator-chart.dhall
time                 253.1 ms   (229.9 ms .. 277.6 ms)
                     0.997 R²   (0.987 R² .. 1.000 R²)
mean                 244.3 ms   (237.1 ms .. 250.3 ms)
std dev              9.151 ms   (6.878 ms .. 10.27 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/deployment.dhall
time                 222.0 ms   (208.5 ms .. 232.3 ms)
                     0.997 R²   (0.990 R² .. 1.000 R²)
mean                 237.9 ms   (229.0 ms .. 250.7 ms)
std dev              13.30 ms   (9.459 ms .. 16.75 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/deploymentSimple.dhall
time                 232.8 ms   (221.5 ms .. 245.4 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 232.3 ms   (227.2 ms .. 235.8 ms)
std dev              5.981 ms   (3.494 ms .. 8.884 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/ingress.dhall
time                 226.1 ms   (217.7 ms .. 234.4 ms)
                     0.998 R²   (0.993 R² .. 1.000 R²)
mean                 228.0 ms   (223.6 ms .. 233.8 ms)
std dev              7.046 ms   (3.615 ms .. 10.28 ms)
variance introduced by outliers: 14% (moderately inflated)

benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/out
bench: ShellFailed {shellCommandLine = "dhall-to-yaml --file ../dhall-kubernetes/examples/out", shellExitCode = ExitFailure 1}
benchmarking dhall-to-yaml --file ../dhall-kubernetes/examples/service.dhall
time                 224.3 ms   (206.8 ms .. 234.3 ms)
                     0.997 R²   (0.986 R² .. 1.000 R²)
mean                 224.0 ms   (218.0 ms .. 227.1 ms)
std dev              5.803 ms   (2.279 ms .. 8.859 ms)
variance introduced by outliers: 14% (moderately inflated)
```

the latest results are better! Thanks for that suggestion @sjakobi @Gabriel439

sjakobi commented on dhall/src/Dhall/Parser/Expression.hs:
```diff
@@ -474,13 +474,12 @@ parsers embedded = Parsers {..}
+                _ <- optional $ try (whitespace *> _comma)
 
-                a <- recordTypeOrLiteral
-
-                whitespace
+                src0 <- src whitespace
```
This still doesn't seem ideal. I think we want to

1. Parse whitespace
2. Check whether there's a comma
3. If there's a comma:
   a) throw away the whitespace from step 1, and parse more whitespace which we'll pass to `rTOL`
   b) otherwise pass the whitespace from step 1 directly to `rTOL`

I think it would look somewhat like this

```haskell
src0 <- whitespace
mComma <- optional _comma
src1 <- case mComma of
    Just _ -> src whitespace
    Nothing -> return src0
a <- recordTypeOrLiteral src1
```

german1608 replied:
Really clever!

I've tested that change and it works. Since it's not really obvious the reason of doing that for future new-comer developers, I documented it properly.

sjakobi commented on dhall/src/Dhall/Parser/Expression.hs:
```diff
@@ -763,22 +767,27 @@ parsers embedded = Parsers {..}
     emptyRecordType = return (Record mempty)
 
-    nonEmptyRecordTypeOrLiteral = do
+    -- Reason of using `try` here is analogue to `emptyRecordLiteral`
```
Is this comment outdated?

german1608 replied:
Yes, I removed it.

german1608 commented:
Thanks @sjakobi and @Gabriel439 for your help and patience!
